Implement DataHub Catalogue client for dp-catalogue library

tom-webber opened the pull request:
- Convert CatalogueClient class into ABC base, so OMD and DataHub client classes can be inherited
- implement `__init__` and `upsert_table` for DataHubCatalogueClient using DataHub gms
- add tests to check DataHub client is submitting requests as expected (in the format supported by DataHub python sdk gms emitter).
- add packages to pypoetry.toml
- refactoring

## Further changes - @MatMoore 
- get tests working with a snapshot testing approach (based on DataHub unit tests)
- add integration test for datahub server
- further refactoring to make the interface catalogue-agnostic
- allow the dataset to be created within a container for its database (e.g. my_data_product_v1)
- remove dummy data from platform schema
- update readme & changelog

Commit:
Convert CatalogueClient class into ABC base, so OMD and DataHub client classes can be inherited
- implement `__init__` and `create_or_update_table` for DataHubCatalogueClient using DataHub gms
- add tests to check DataHub client is submitting requests as expected (in the format supported by DataHub python sdk gms emitter).
- add packages to pypoetry.toml
- refactoring

Commit:
- remove unused `get_client_by_name` method for fetching client based on name or enum
- clear up datahub docstring/comments

Commit:
remove test file

Commit:
fix test integration

Commit:
fix linting

Commit:
mypy lint fix

Commit:
mypy linting

Commit:
as above

Commit:
more mypy ignore comments

PriyaBasker23 commented on python-libraries/data-platform-catalogue/data_platform_catalogue/client.py:
```diff
@@ -79,13 +178,13 @@ class CatalogueClient:
         self,
         jwt_token,
-        api_uri: str = "https://catalogue.apps-tools.development.data-platform.service.justice.gov.uk/api",
+        api_url: str = "https://catalogue.apps-tools.development.data-platform.service.justice.gov.uk/api",
```
Can this be moved to config file

jemnery replied:
Is this a config which sets env vars? Environment variables could make it easier to deploy a different `api_url` (etc) to different environments via GHA in future.

tom-webber replied:
This is just a default argument. The actual value used for client setup would be from usage of this catalogue package, and could be set by an env value.

tom-webber replied:
default string has been removed.

PriyaBasker23 commented on python-libraries/data-platform-catalogue/data_platform_catalogue/client.py:
```diff
@@ -201,47 +307,125 @@ def get_user_id(self, user_email: str):
+    def __init__(
+        self,
+        jwt_token,
+        api_url: str = "https://datahub.apps-tools.development.data-platform.service.justice.gov.uk/api/gms",
```
Can this be moved to config file

tom-webber replied:
default string has been removed.

Commit:
update DataHubCatalogueClient documentation
- handle misconfigured api_urls in DataHub client
- make tags api call optional if not present
- update tests for changes

Commit:
lint

tom-webber committed:
Merge branch 'main' into add-datahub-client-to-dp-catalogue-pkg

Commit:
linting?

Commit:
mypy linter driving me nuts

Commit:
test mypy definition line

MatMoore commented on python-libraries/data-platform-catalogue/data_platform_catalogue/client.py:
```diff
@@ -382,17 +395,36 @@ def create_or_update_schema(self, metadata: DataProductMetadata, database_fqn: s
+            platform=make_data_platform_urn(platform=platform),
+            version=version,
             hash="",
             platformSchema=OtherSchemaClass(rawSchema="__insert raw schema here__"),
```
this field is still a placeholder so I think we should either remove it (if not needed) or populate it

PriyaBasker23 replied:
yes , was hoping to mention the same

PriyaBasker23 replied:
what is  hash="" - used for ?. Can we remove any parameters if its not useful

tom-webber replied:
`platformSchema` is not an optional field, and needs to be populated. Unfortunately I'm not sure _what_ it needs to be populated _with_.

Same for `hash`.

MatMoore replied:
hash is "the SHA1 hash of the schema content" and platformSchema is "The native schema in the dataset's platform." according to https://datahubproject.io/docs/generated/metamodel/entities/dataset/#schemametadata

Don't know what we would actually pass though, or if it matters that much. We already have the native types passed through for each field, so maybe these can both be left as empty strings?

Commit:
- add `create_domain` and `create_or_update_data_product` methods to DataHubCatalogueClient class
- update `DataHubCatalogueClient.create_or_update_table` method to create domain and data product if they don't exist but are passed as `data_product_metadata`
- associate tables with data products when created in DataHub

Commit:
rename all `create_or_update_x` methods to `upsert_x`

Commit:
fix rename in tests
linting

Commit:
putting in pre-requisites for using MockDataHubGraph as per https://github.com/datahub-project/datahub/blob/master/metadata-ingestion/tests/unit/api/entities/dataproducts/test_dataproduct.py#L10C46-L10C62

MatMoore committed:
Add missing dependency for datahub testing code

MatMoore committed:
Add missing code for golden file testing

Taken from datahub's testing code
https://github.com/datahub-project/datahub/blob/master/metadata-ingestion/tests/unit/api/entities/dataproducts/test_dataproduct.py#L10C46-L10C62

MatMoore committed:
Add starting point for golden file testing

From Datahub codebase

MatMoore committed:
Ensure table name is consistent

MatMoore committed:
Delint

MatMoore committed:
More linting

MatMoore committed:
Remove unused testing code

MatMoore committed:
Change the datahub create table test to a snapshot test

MatMoore committed:
Lint

MatMoore committed:
Ignore lint in generated files

MatMoore committed:
Don't complain about line endings on snapshot files

MatMoore committed:
Test cases where we create data product, domain, table at the same time

MatMoore committed:
Remove tests of underlying library functionality

MatMoore committed:
Split tests into two modules

MatMoore committed:
Refactor test

MatMoore commented on python-libraries/data-platform-catalogue/data_platform_catalogue/client.py:
```diff
@@ -201,47 +310,277 @@ def get_user_id(self, user_email: str):
+            entityUrn=dataset_urn,
+            aspect=dataset_schema_properties,
+        )
+        self.graph.emit(metadata_event)
```
⚠️ From integration testing this, it looks like the `SchemaPropertiesClass` we pass here is not enough for the dataset to show up in the UI. I *think* we need a DatasetPropertiesClass aspect as well, i.e. this snippet

```python
dataset_properties = DatasetPropertiesClass(
    description="Example table for testing",
    customProperties={"dpia_required": "yes"},
)

# Construct a MetadataChangeProposalWrapper object.
metadata_event = MetadataChangeProposalWrapper(
    entityUrn=builder.make_dataset_urn(
        "athena", "athena.catalog.hmpps_use_of_force_v2.statements"
    ),
    aspect=dataset_properties,
)
```

MatMoore commented on python-libraries/data-platform-catalogue/data_platform_catalogue/client.py:
```diff
@@ -65,9 +121,55 @@ class ReferencedEntityMissing(CatalogueError):
+        pass
+
+    @abstractmethod
+    def upsert_table(
```
I don't think the use of `*args` and `**kwargs` is doing what you expect here.

Since this is an abstract base class, we would expect to be able to type any value to `BaseCatalogueClient` and then be able to substitute any of the concrete classes without breaking things. But this is not the case, since the base class accepts *any* arguments, and the concrete classes do not.

(See [Liskov's substitution principle](https://en.wikipedia.org/wiki/Liskov_substitution_principle))

I think what you want to do instead is write out the full arglist here, and then use *args, **kwargs in the subclasses only when you want to ignore some of them.

MatMoore committed:
Split client module into smaller files

MatMoore committed:
Remove unused enum

MatMoore committed:
Add DatasetProperties aspect

This is needed in addition to the SchemaMetadata aspect

MatMoore commented:
I have tested using this branch to create a dataset on the apps and tools instance. See https://datahub.apps-tools.development.data-platform.service.justice.gov.uk/dataset/urn:li:dataset:(urn:li:dataPlatform:glue,my_data_product2.my_table3,PROD)/Schema?is_lineage_mode=false&schemaFilter=

I think I'm reasonably confident it's working end to end now, although there are some things we might need to come back to.

Notes:

- Custom variables are not implemented yet - so expecting we'll need to add these and any other missing fields later
- We have to pass tags at the table level now, since DatahubCatalogueClient is not passing along any tags from the Data Product (and either way I don't believe the dataset would inherit this from the data product).
- <strike>With the DatahubCatalogueClient, you need to pass the name as $database_name.$table_name, if you want the database to show up as a container. Otherwise it sticks everything in a "Default" container.</strike> This is no longer necessary
- In datahub entities have multiple "aspects". We were creating an aspect for the schema, and tags, but we were missing the one for basic properties associated with the dataset, so I've added that in.

MatMoore committed:
De-linting continues

MatMoore committed:
lint

MatMoore committed:
Ensure subtypes conform to base type

The aim here is to hide catalogue specific details from the user of
this library.

This means we should expect to be able to type any value to
BaseCatalogueClient and then be able to substitute any of the concrete
classes without breaking things. (See Liskov's substitution principle)

This cannot be the case if the base class has methods which accept *any*
arguments, and the concrete implementations only handle specific
arguments.

To fix this I've modified the abstract methods to specify the arguments.

Since the schema_fqn, database_fqn, etc arguments reflected the
OpenMetadata graph structure, I've replaced this with a generic
"DataLocation" argument, which does not enforce a particularly
hierarchy.

MatMoore committed:
lint

MatMoore commented on python-libraries/data-platform-catalogue/data_platform_catalogue/client/openmetadata.py:
```diff
@@ -201,42 +210,22 @@ def get_user_id(self, user_email: str):
+            raise ReferencedEntityMissing
 
     def _generate_tags(self, tags: list[str]):
+        # TODO? update using the sdk logic:
```
There is an unaddressed TODO here which I'm just going to leave since it's the OpenMetadata implementation, and we don't know if we'll be returning to this.

MatMoore committed:
lint

MatMoore committed:
Tweak datahub output

- Pass a fully qualified table name ('database.table' rather than 'table') so
  that a container is created rather than allocating to the default
  container.

- Remove dummy value for mandatory platformSchema argument. Just use
  empty string.

MatMoore commented:
Here is a table created with the latest code:

https://datahub.apps-tools.development.data-platform.service.justice.gov.uk/dataset/urn:li:dataset:(urn:li:dataPlatform:glue,test_data_product_v2.test_table,PROD)/Schema?is_lineage_mode=false&schemaFilter=

Note that I have fully qualified the database name here (database.table), which groups everything from the same database in a container that shows up in the breadcrumb, and in the datahub search interface.
![Screenshot 2024-01-18 at 15 25 36](https://github.com/ministryofjustice/data-platform/assets/87579/cd699717-8c3e-4a71-874f-8b584c334eb1)

We can change this later if we don't like it, but this keeps it consistent with what we were doing before, in the OpenMetadata implementation.

PriyaBasker23 commented:
> Here is a table created with the latest code:
> 
> https://datahub.apps-tools.development.data-platform.service.justice.gov.uk/dataset/urn:li:dataset:(urn:li:dataPlatform:glue,test_data_product_v2.test_table,PROD)/Schema?is_lineage_mode=false&schemaFilter=
> 
> Note that I have fully qualified the database name here (database.table), which groups everything from the same database in a container that shows up in the breadcrumb, and in the datahub search interface. ![Screenshot 2024-01-18 at 15 25 36](https://private-user-images.githubusercontent.com/87579/297793719-cd699717-8c3e-4a71-874f-8b584c334eb1.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDU2NzExNjMsIm5iZiI6MTcwNTY3MDg2MywicGF0aCI6Ii84NzU3OS8yOTc3OTM3MTktY2Q2OTk3MTctOGMzZS00YTcxLTg3NGYtOGI1ODRjMzM0ZWIxLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAxMTklMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMTE5VDEzMjc0M1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWM3MjJmOTZlNTg1YjQ3MjQ5ZjI4MjQzZWQ0OTFhNzZlMTE2OTY5ZTk1NjgzYWQ4NDE1NTc4ZDE0Yjk5MzJmYzMmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.XoMzjOsF4n5S_X1AWtOpEU3qH_9b2tRpfe-F1VKNz6M)
> 
> We can change this later if we don't like it, but this keeps it consistent with what we were doing before, in the OpenMetadata implementation.

It would be good to map tables into its own database

julialawrence reviewed (approved):
Priya approved.
